Be more strict with access resource titles / ACL syntax

smortex opened the pull request:
The openldap_access resource allows a lot of variations in the title for
declaring a resource, making it possible to skip passing parameters such
as `what` and `suffix`.  This flexibility however can confuse Puppet
when it is prefetching resources, leading to catalog compilation
failures.

Impose a more strict format for resource titles, and validate it with
tighter custom types to raise a hopefully meaningful error instead of a
Ruby error because something borked bad.

Fixes #294

bastelfreak commented on manifests/server/access.pp:
```diff
@@ -1,23 +1,27 @@
   include openldap::server
 
+  if $position or $suffix {
+    warning('openldap::server::access position and suffix are deprecated')
```
should we also add a notify here? warnings are only logged to the server, not on the agent side.

smortex replied:
Yes, a notify would be better suited.

Now I am wondering if we should better just drop this since that mean for sure 2 breaking releases for access fixes, which have already introduced major releases in the past?

neomilium replied:
IMHO : `Array[Hash[String[1], Any]]` == :+1:

smortex replied:
Thanks, I updated the PR accordingly.

smortex commented on manifests/server/access.pp:
```diff
@@ -1,23 +1,27 @@
+  Array[Openldap::Access_rule]         $access,
+  Enum['present', 'absent']            $ensure   = 'present',
+  Optional[Variant[Integer,String[1]]] $position = undef, # FIXME deprecated
+  Optional[String[1]]                  $suffix   = undef, # FIXME deprecated
```
Good catch. I was feeling bad because I am afraid of meeting this guy in the street:

![I am once again asking you stop breaking the API](https://user-images.githubusercontent.com/148721/136709998-2293acab-5294-49bc-afc7-c4aca42e84a9.png)

smortex replied:
The next release being a major one, I decided to remove the legacy code completely and added a commit for this.

smortex replied:
The commit has been squashed for the sake of simplicity :wink:

bastelfreak commented on types/access_title.pp:
```diff
@@ -0,0 +1,2 @@
+# A valid title for an openldap::server::access resource
```
can you update this as puppet-strings style?
```suggestion
# @summary A valid title for an openldap::server::access resource
```

bastelfreak commented on types/access_rule.pp:
```diff
@@ -0,0 +1,2 @@
+# A valid access rule for openldap::server::access
```
```suggestion
# @summary A valid access rule for openldap::server::access
```

bastelfreak commented on types/access_hash.pp:
```diff
@@ -0,0 +1,10 @@
+# A valid acl value for openldap::server::access_wrapper
```
```suggestion
# @summary A valid acl value for openldap::server::access_wrapper
```

smortex commented:
I rebased these changes on top of master, squashed the minor fixes that where added part of previous reviews, and added a new commit to address the final issue that worried me with this change.

Can I have a review please?

bastelfreak commented on manifests/server/access_wrapper.pp:
```diff
@@ -10,35 +10,64 @@
 #
 # [*acl*]
```
we should also update this to puppet-strings soonish

smortex replied:
I saw this yesterday and bravely decided to go look away :laughing:
